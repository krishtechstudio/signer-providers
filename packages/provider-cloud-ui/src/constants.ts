import { BigNumber } from '@waves/bignumber';
import { TransactionType } from '@waves/ts-types';

export const MAX_ALIAS_LENGTH = 30;

export const WAVES = {
    ticker: 'WAVES',
    assetId: 'WAVES',
    name: 'Waves',
    decimals: 8,
    description: '',
    issueHeight: 0,
    issueTimestamp: new Date('2016-04-11T21:00:00.000Z').getTime(),
    reissuable: false,
    scripted: false,
    minSponsoredFee: null,
    quantity: BigNumber.toBigNumber(100000000).mul(Math.pow(10, 8)).toFixed(),
    issuer: 'WAVES',
    minSponsoredAssetFee: null,
    logo: '',
};

export const NAME_MAP = {
    issue: 3 as const,
    transfer: 4 as const,
    reissue: 5 as const,
    burn: 6 as const,
    exchange: 7 as const,
    lease: 8 as const,
    cancelLease: 9 as const,
    alias: 10 as const,
    massTransfer: 11 as const,
    data: 12 as const,
    setScript: 13 as const,
    sponsorship: 14 as const,
    setAssetScript: 15 as const,
    invoke: 16 as const,
    updateAssetInfo: 17 as const,
};

export const SPONSORED_TYPES: Array<TransactionType> = [
    NAME_MAP.transfer,
    NAME_MAP.invoke,
];
