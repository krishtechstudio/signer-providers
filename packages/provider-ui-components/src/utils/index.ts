export * from './analytics';
export * from './assetPropFactory';
export * from './cleanAlias';
export * from './fixRecipient';
export * from './getEnvAwareUrl';
export * from './getTxAliases';
export * from './isAddress';
export * from './isAlias';
export * from './loadFeeByTransaction';
export * from './loadLogoInfo';
export * from './math';
export * from './Queue';
export * from './toArray';
export * from './toFormat';
