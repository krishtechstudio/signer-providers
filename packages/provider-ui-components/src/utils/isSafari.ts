export function isSafari(): boolean {
    const userAgent = navigator.userAgent.toLowerCase();
    const isSafariUA =
        userAgent.includes('safari') && !userAgent.includes('chrome');
    const iOS =
        navigator.platform != null &&
        /iPad|iPhone|iPod/.test(navigator.platform);

    return iOS || isSafariUA;
}

export function isBrave(): boolean {
    return !!(navigator as any).brave?.isBrave;
}
