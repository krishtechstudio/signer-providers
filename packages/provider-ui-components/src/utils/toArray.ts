export const toArray = <T>(data: T | Array<T>): Array<T> =>
    Array.isArray(data) ? data : [data];
