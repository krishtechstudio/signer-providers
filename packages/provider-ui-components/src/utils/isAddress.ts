import { cleanAddress } from './cleanAlias';
import { MAX_ALIAS_LENGTH } from '../constants';

export function isAddress(address: string): boolean {
    return cleanAddress(address).length > MAX_ALIAS_LENGTH;
}
