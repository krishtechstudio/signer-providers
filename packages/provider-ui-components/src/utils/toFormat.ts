import { TAssetDetails } from '@waves/node-api-js/es/api-node/assets';
import { getPrintableNumber } from './math';
import { Long } from '@waves/ts-types';
import { WAVES } from '../constants';

export function toFormat(
    num: Long,
    id: string | null,
    hash: Record<string, TAssetDetails>
): string {
    const asset = id != null ? hash[id] : WAVES;

    if (asset == null) {
        throw new Error('Asset not found!');
    }

    return `${getPrintableNumber(num, asset.decimals)} ${asset.name}`;
}
