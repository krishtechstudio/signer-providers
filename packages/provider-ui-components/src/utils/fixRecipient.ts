import { curry } from 'ramda';
import { isAddress } from './isAddress';
import { cleanAddress } from './cleanAlias';

export const fixRecipient = curry(
    <T extends { recipient: string }>(networkByte: number, data: T): T => {
        const address = cleanAddress(data.recipient);

        if (isAddress(address)) {
            return { ...data, recipient: address };
        } else {
            return {
                ...data,
                recipient: `alias:${String.fromCharCode(
                    networkByte
                )}:${address}`,
            };
        }
    }
);
