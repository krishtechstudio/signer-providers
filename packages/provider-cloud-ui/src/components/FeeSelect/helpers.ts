import { FeeOption } from '@waves.exchange/react-uikit';
import { BigNumber } from '@waves/bignumber';
import { Long } from '@waves/ts-types';
import { DetailsWithLogo } from '../../interface';
import { WAVES } from '../../constants';
import { assetPropFactory } from '../../utils/assetPropFactory';
import { getPrintableNumber } from '../../utils/math';
import { FeeSelectTxMeta } from './FeeSelect';

export const checkIsEnoughBalance = (balance: Long, fee: Long): boolean => {
    return BigNumber.toBigNumber(balance).gte(
        BigNumber.toBigNumber(fee).div(Math.pow(10, WAVES.decimals))
    );
};

const hasParamsFee = (fee: Long | undefined): fee is Long =>
    typeof fee === 'string' || typeof fee === 'number';

const isFeeAssetId = (
    feeAssetId: string | null | undefined
): feeAssetId is string | null => typeof feeAssetId !== 'undefined';

const isNonDefaultFeeAssetId = (
    feeAssetId: string | null | undefined
): feeAssetId is string => typeof feeAssetId === 'string';

export const formatFee = (fee: Long, decimals: number): string =>
    getPrintableNumber(fee, decimals);

type GetFeeOptions = (arg: {
    txFee: Long;
    paramsFee: Long | undefined;
    txMeta: FeeSelectTxMeta;
    paramsFeeAssetId: string | null | undefined;
    availableWavesBalance: Long;
}) => FeeOption[];

export const getFeeOptions: GetFeeOptions = ({
    txFee,
    txMeta,
    paramsFeeAssetId,
    availableWavesBalance,
    paramsFee: txParamsFee,
}) => {
    const getAssetProp = assetPropFactory(txMeta.assets);
    const wavesFeeOption: FeeOption = {
        id: null,
        name: WAVES.name,
        ticker: WAVES.ticker,
        value: formatFee(txFee, WAVES.decimals),
    };

    let defaultFeeOption: FeeOption;
    let feeAsset: DetailsWithLogo;

    if (isFeeAssetId(paramsFeeAssetId) || hasParamsFee(txParamsFee)) {
        if (isNonDefaultFeeAssetId(paramsFeeAssetId)) {
            // case feeAssetId - some asset, but not Waves, fee is not provided
            feeAsset = txMeta.assets[paramsFeeAssetId];
            defaultFeeOption = {
                name: getAssetProp(paramsFeeAssetId, 'name'),
                id: paramsFeeAssetId,
                ticker: '',
                value: formatFee(txFee, feeAsset.decimals),
            };
        } else {
            // case: feeAssetId - Waves or fee provided
            defaultFeeOption = wavesFeeOption;
        }

        // in case feeAssetId provided we can only use it
        return [defaultFeeOption];
    }
    // case: no feeAssetId provided
    // feeAsset = WAVES;
    defaultFeeOption = wavesFeeOption;

    const metaFeeOptions = txMeta.feeList.map((f) => ({
        name: getAssetProp(f.feeAssetId, 'name'),
        id: String(f.feeAssetId),
        ticker: '',
        value: formatFee(f.feeAmount, getAssetProp(f.feeAssetId, 'decimals')),
    }));

    const isEnoughBalance = checkIsEnoughBalance(availableWavesBalance, txFee);
    const hasNonDefaultFees = metaFeeOptions.length > 0;

    let feeOptions;

    if (hasNonDefaultFees) {
        feeOptions = isEnoughBalance
            ? [defaultFeeOption].concat(metaFeeOptions)
            : metaFeeOptions.filter(({ id }) => id !== WAVES.assetId);
    } else {
        feeOptions = [defaultFeeOption];
    }

    return feeOptions;
};
