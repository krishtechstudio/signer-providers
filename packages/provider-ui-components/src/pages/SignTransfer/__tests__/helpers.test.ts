import { AssetDecimals, Long, MassTransferItem } from '@waves/ts-types';
import { NAME_MAP, WAVES } from '../../../constants';
import * as math from '../../../utils/math';
import * as helpers from '../helpers';

const massTransferTxType = NAME_MAP.massTransfer;
const transferTxType = NAME_MAP.transfer;
const asset1Id = 'asset1';
const asset1Name = 'asset1_name';
const asset1Decimals = 2 as AssetDecimals;
const asset1 = {
    assetId: asset1Id,
    decimals: asset1Decimals,
    name: asset1Name,
};
const assets: helpers.MetaAssets = {
    [asset1Id]: asset1,
};
const actualAddress = '3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj';
const alias1 = 'alias:T:merry';
const alias1Address = 'alias1_address';
const aliases: helpers.MetaAliases = {
    [alias1]: alias1Address,
};
const massTransferItem1: MassTransferItem<Long> = {
    amount: 1,
    recipient: actualAddress,
};
const massTransferItem2: MassTransferItem<Long> = {
    amount: 2,
    recipient: alias1,
};

describe('getAmountAsset', () => {
    it('returns WAVES if assetId === null', () => {
        expect(helpers.getAmountAsset(null, assets)).toEqual(WAVES);
    });
    it('returns WAVES if assetId === undefined', () => {
        expect(helpers.getAmountAsset(undefined, assets)).toEqual(WAVES);
    });
    it('returns correct asset if non waves asset id passed', () => {
        expect(helpers.getAmountAsset(asset1Id, assets)).toEqual(asset1);
    });
});

describe('getAssetName', () => {
    it('returns WAVES asset name if assetId === null', () => {
        expect(helpers.getAssetName(null, assets)).toBe(WAVES.name);
    });
    it('returns WAVES asset name if assetId === undefined', () => {
        expect(helpers.getAssetName(undefined, assets)).toBe(WAVES.name);
    });
    it('returns correct asset name if non waves asset id passed', () => {
        expect(helpers.getAssetName(asset1Id, assets)).toBe(asset1.name);
    });
});

describe('getFeeAsset', () => {
    it('returns WAVES if mass transfer', () => {
        expect(helpers.getFeeAsset(11, assets)).toEqual(WAVES);
    });
    it('returns WAVES if transaciton fee asset id is undefined', () => {
        expect(helpers.getFeeAsset(4, assets)).toEqual(WAVES);
    });
    it('returns WAVES if transaciton fee asset id === Waves id', () => {
        expect(helpers.getFeeAsset(4, assets, null)).toEqual(WAVES);
    });
    it('returns correct asset if non waves fee asset id passed', () => {
        expect(helpers.getFeeAsset(4, assets, asset1Id)).toEqual(asset1);
    });
});

describe('getRecipientAddress', () => {
    it('returns address for alias', () => {
        expect(helpers.getRecipientAddress(alias1, aliases)).toEqual(
            aliases[alias1]
        );
    });
    it('returns address', () => {
        expect(helpers.getRecipientAddress(actualAddress, aliases)).toEqual(
            actualAddress
        );
    });
});

describe('getRawTransfersList', () => {
    it('returns correct transfers list', () => {
        expect(
            helpers.getRawTransfersList(aliases, [
                massTransferItem1,
                massTransferItem2,
            ])
        ).toMatchInlineSnapshot(`
            Array [
              Object {
                "address": "3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj",
                "amount": 1,
                "name": "3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj",
              },
              Object {
                "address": "alias1_address",
                "amount": 2,
                "name": "alias:T:merry",
              },
            ]
        `);
    });
});

describe('getTotalTransferAmount', () => {
    const rawTransferList = [
        {
            address: '3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj',
            amount: 1,
            name: '3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj',
        },
        {
            address: 'alias1_address',
            amount: 2,
            name: 'alias:T:merry',
        },
    ];

    it('calculates correct total amount #1', () => {
        const decimals = 2;
        const actual = helpers.getTotalTransferAmount(
            rawTransferList,
            decimals
        );

        expect(actual).toBe('0.03');
    });
    it('calculates correct total amount #2', () => {
        const decimals = 4;
        const actual = helpers.getTotalTransferAmount(
            rawTransferList,
            decimals
        );

        expect(actual).toBe('0.0003');
    });
});

describe('getTransferList', () => {
    const rawTransferList = [
        {
            address: '3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj',
            amount: 1,
            name: '3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj',
        },
        {
            address: 'alias1_address',
            amount: 2,
            name: 'alias:T:merry',
        },
    ];
    const decimals = 2;

    let getPrintableNumberMock: any;

    beforeAll(() => {
        getPrintableNumberMock = jest
            .spyOn(math, 'getPrintableNumber')
            .mockImplementation(jest.fn((amount) => String(amount)));
    });

    afterEach(() => {
        getPrintableNumberMock.mockReset();
    });
    afterEach(() => {
        getPrintableNumberMock.mockRestore();
    });

    it('returns correct transfer list', () => {
        const actual = helpers.getTransferList(rawTransferList, decimals);

        expect(getPrintableNumberMock).toHaveBeenCalledTimes(2);
        expect(getPrintableNumberMock).toHaveBeenNthCalledWith(
            1,
            rawTransferList[0].amount,
            decimals
        );
        expect(getPrintableNumberMock).toHaveBeenNthCalledWith(
            2,
            rawTransferList[1].amount,
            decimals
        );
        expect(actual).toMatchSnapshot();
    });
});

describe('getFeeAssetName', () => {
    describe('returns WAVES name if', () => {
        it('mass transfer', () => {
            expect(helpers.getFeeAssetName(11, assets)).toBe(WAVES.name);
        });
        it('transaciton fee asset id is undefined', () => {
            expect(helpers.getFeeAssetName(4, assets)).toEqual(WAVES.name);
        });
        it('fee asset id === Waves id', () => {
            expect(helpers.getFeeAssetName(4, assets, null)).toEqual(
                WAVES.name
            );
        });
    });
    it('returns correct asset if non waves fee asset id passed', () => {
        expect(helpers.getFeeAssetName(4, assets, asset1Id)).toEqual(
            asset1.name
        );
    });
});

describe('getPrintableTxFee', () => {
    it('returns correct value #1', () => {
        const txFee = '100000000';
        const txType = massTransferTxType;
        const actual = helpers.getPrintableTxFee({
            txType,
            txFee,
            assets,
        });

        expect(actual).toBe(`1 ${WAVES.name}`);
    });

    it('returns correct value #2', () => {
        const txFee = '100';
        const txType = transferTxType;
        const txFeeAssetId = asset1Id;

        const actual = helpers.getPrintableTxFee({
            txType,
            txFee,
            assets,
            txFeeAssetId,
        });

        expect(actual).toBe(`1 ${asset1Name}`);
    });
});

describe('getTransferViewData', () => {
    it('returns correct value', () => {
        const txRecipient = alias1;
        const txAssetId = asset1Id;
        const txFee = '100000000';
        const txFeeAssetId = null; //Waves
        const txAmount = '100000000';

        const actual = helpers.getTransferViewData({
            txRecipient,
            txAssetId,
            txFee,
            txFeeAssetId,
            txAmount,
            assets,
            aliases,
        });

        expect(actual).toMatchSnapshot();
    });
});

describe('getMassTransferViewData', () => {
    it('returns correct value', () => {
        const txAssetId = asset1Id;
        const txFee = '100000000';
        const txTransfers = [massTransferItem1, massTransferItem2];

        const actual = helpers.getMassTransferViewData({
            txTransfers,
            txAssetId,
            txFee,
            assets,
            aliases,
        });

        expect(actual).toMatchSnapshot();
    });
});

describe('getViewData', () => {
    beforeAll(() => {
        jest.spyOn(helpers, 'getTransferViewData').mockImplementation(
            () => ({} as any)
        );
        jest.spyOn(helpers, 'getMassTransferViewData').mockImplementation(
            () => ({} as any)
        );
    });
    it("returns no attachement if tx has't one", () => {
        const actual = helpers.getViewData({} as any, {} as any);

        expect(actual.attachment).toBeFalsy();
    });

    it('returns correct attachment', () => {
        const attachment = '72k1xXWG59fYdzSNoA';
        const actual = helpers.getViewData({ attachment } as any, {} as any);

        expect(actual.attachment).toBe('Hello, World!');
    });
});
