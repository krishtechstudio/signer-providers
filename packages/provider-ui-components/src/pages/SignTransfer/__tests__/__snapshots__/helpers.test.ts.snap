// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMassTransferViewData returns correct value 1`] = `
Object {
  "fee": "1 Waves",
  "totalTransferAmount": "0.03 asset1_name",
  "transferList": Array [
    Object {
      "address": "3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj",
      "amount": "0.01",
      "name": "3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj",
    },
    Object {
      "address": "alias1_address",
      "amount": "0.02",
      "name": "alias:T:merry",
    },
  ],
}
`;

exports[`getTransferList returns correct transfer list 1`] = `
Array [
  Object {
    "address": "3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj",
    "amount": "1",
    "name": "3PCAB4sHXgvtu5NPoen6EXR5yaNbvsEA8Fj",
  },
  Object {
    "address": "alias1_address",
    "amount": "2",
    "name": "alias:T:merry",
  },
]
`;

exports[`getTransferViewData returns correct value 1`] = `
Object {
  "fee": "1 Waves",
  "totalTransferAmount": "1000000 asset1_name",
  "transferList": Array [
    Object {
      "address": "alias1_address",
      "amount": "1000000",
      "name": "alias:T:merry",
    },
  ],
}
`;
